RDP client implementation

awly opened the pull request:
This is a minimal RDP client on top of the rdp-rs Rust crate.
The crate is wrapped with an FFI Rust adapter, statically compiled and
called from Go via CGO.

This PR also has toy web client to test the RDP code, and just enough of
the desktop wire protocol implementation to make it work.

I excluded the RDP client from the build via tags for now to avoid
bloating the teleport binaries.

Many more things missing that will come in later PRs, to keep this one
reviewable.

@fspmarshall and @xacrimon this is my first "real" Rust code, feel free to tear it apart!

Commit:
RDP client implementation

This is a minimal RDP client on top of the rdp-rs Rust crate.
The crate is wrapped with an FFI Rust adapter, statically compiled and
called from Go via CGO.

This PR also has toy web client to test the RDP code, and just enough of
the desktop wire protocol implementation to make it work.

I excluded the RDP client from the build via tags for now to avoid
bloating the teleport binaries.

Many more things missing that will come in later PRs, to keep this one
reviewable.

fspmarshall reviewed (commented):
Did an initial pass for low-hanging memory issues. Will need to circle back.

There are various issues with how the `CGOString` value is being used:
- Conversion to rust string assumes pointer is UTF-8, when this may not be true.  Non UTF-8 values can cause memory unsafety in rust strings, so this invariant must hold.
- Memory allocated in go is sometimes freed in rust, and vise-versa (also there are a few double-frees).  `C.free` is not compatible with values allocated by rust, and rust's `Drop` trait is not compatible with values allocated by cgo.
- Rust strings are created with potentially incorrect capacity values, which is generally unsafe.


...luckily there is a fairly straightforward solution to these problems.  Since rust and go both provide good tools for working with regular null-terminated C strings, its best to simply do away with the custom `CGOString` type entirely.  Use rust's [`CStr`](https://doc.rust-lang.org/std/ffi/struct.CStr.html) and [`CString`](https://doc.rust-lang.org/std/ffi/struct.CString.html) types to represent go-allocated and rust-allocated strings respectively in your rust code.  The docs for those types explain their usage better than I can, but its fairly straightforward.  `CStr` is a "borrowed" value and not necessarily owned by rust, and `CString` is an "owned" value that was allocated by rust.  These types provide safe methods and trait impls for converting to and from rust's string types, and interact correctly with the borrow checker.

I suggest taking a look at [this comment](https://github.com/gravitational/teleport/pull/7543#discussion_r669187769) for an example of correct memory handling on the go side.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -0,0 +1,308 @@
+func (c *Client) handleBitmap(cb C.CGOBitmap) C.CGOError {
+	data := C.GoBytes(unsafe.Pointer(cb.data_ptr), C.int(cb.data_len))
+	// Convert BGRA to RGBA. This is probably some endianness weirdness, if
+	// Windows handles pixels in ARGB as uint32.
```
Could it be a that network formats almost always use big-endian perhaps?

xacrimon replied:
Should probably include a comment about that. Looks solid otherwise.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+
+// Rust-side registry of clients, to allow Go to reference a specific client in calls after
+// connect_rdp.
+lazy_static! {
```
Generally globals are heavily discouraged. Does it make more sense to have a Rust object with a pointer passed around C GUI lib style? It can of course make sense to have this but it should be carefully considered.

awly replied:
Yeah, this feels gross.
I couldn't figure out how to toss a Rust object pointer over to the Go side, while keeping the object from being cleaned up by Rust's memory management. Also, I'm not sure how to safe it is to reconstruct a Rust object from a raw pointer (all the examples at https://doc.rust-lang.org/1.30.0/book/first-edition/raw-pointers.html deal with primitive types).

Do you know of any docs or examples I could follow?

xacrimon replied:
Generally I'd recommend reading the official [nomicon](https://doc.rust-lang.org/nomicon/) as it teaches various dark arts. Rust's memory management is passive and you can use `std::mem::forget` to prevent a struct from being cleaned up. That said the way you want to do this is to `Box<T>` your data object and then call `Box::leak` which as the name suggests, forgets about the memory and gives you a pointer to the value on the heap. You'd then use Go's raw package to deal with this raw pointer. Just don't make it look like a Go value pointer, it needs to be managed as as FFI pointer in Go to prevent it from being GCd. On the Rust side, call `Box::from_raw` to turn the pointer back into a boxed value you can actually use.

xacrimon replied:
The more technical term is that structs that manage heap resources such as `Vec` and `Box` implement the Drop trait which acts as a finalizer that runs at end of scope. `mem::forget` prevents that from being run.

xacrimon replied:
Also, read the entire nomicon before you continue, it takes a couple hours, it's a great read and prevents you from making many mistakes.

xacrimon replied:
That does indeed look correct. But you probably want to give Valgrind, TSAN and MSAN run over this if possible.

xacrimon replied:
The invariant that has to be upheld here is that you cannot call multiple functions on the same object concurrent without horrible UB so you might want a lock on the Go side.

awly replied:
I added an `Arc<Mutex<...>>` on top of the rust RDP client object: https://github.com/gravitational/teleport/blob/82d382de8f9fbffd318094ebaeb0579c1c574541/lib/srv/desktop/rdp/rdpclient/src/lib.rs#L35
The other field (`tcp_fd`) is read-only.

For releasing memory, I split `free_rdp` away from `close_rdp`.
That way the Go client only releases the Rust memory after all i/o goroutines have finished (after `wg.Wait()` returned).

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+    }
+}
+
+fn with_client<F: FnMut(&SyncRdpClient) -> Result<(), ClientError>>(
```
Usually `std::option::Option` is preferred to semantically encode the notion of something potentially not existing while `std::result::Result` is used to encode the notion of an operation expected to succeed going wrong. This to me seems like the former.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+        .insert(client_ref, Arc::new(Mutex::new(client)));
+}
+
+fn unregister_client(client_ref: &i64) {
```
Primitive integers like `i64` should never be taken by immutable reference. This is because the reference is just as big and thus you win no performance and LLVM can't optimize this unless it decides to inline thus function. Generally types which implement the `Copy` trait from `std` are small and designed to be copied instead of immutably referenced.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+
+// CGOString is a CGO-compatible definition of a UTF-8 string that can be used in Go and Rust.
+// This string is not null-terminated.
+#[repr(C)]
```
use `std::ffi::{CString, CStr}` instead

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+// Important: when converting String to CGOString, the caller is responsible for manually freeing
+// the memory at CGOString.data.
+impl From<String> for CGOString {
+    fn from(mut s: String) -> CGOString {
```
Just as a sidenote aside from all the horrible allocator related UB this would inflict as @fspmarshall noted with incompatible allocators, Rust allocations have a "layout" associated with them that have to be used when freeing. I'd recommend reading up on the docs for `std::alloc`.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+// size. If succeeded, the client is internally registered under client_ref. When done with the
+// connection, the caller must call close_rdp.
+#[no_mangle]
+pub extern "C" fn connect_rdp(
```
This function should just be a thin wrapper to a more "rustic" function that utilizes `Result` for errors. This allows you to use the `?` operator for short circuiting error cases which is a lot nicer and foolproof. Generally extern c functions should only deal with conversions and calling another Rust-style function.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+// CGOString is a CGO-compatible definition of a UTF-8 string that can be used in Go and Rust.
+// This string is not null-terminated.
+#[repr(C)]
+pub struct CGOString {
```
The various converters on this also allow you to create UB or segfault in safe Rust. Generally this should be impossible and if your code allows you to do that, you need checked conversions like std provides for its ffi types.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+    }
+}
+
+// TODO: this is Linux-specific, also implement for Windows.
```
You need to conditionally compile this for windows. Add this above the function below the comment.
`#[cfg(target_os = "linux")]`
Read more here https://doc.rust-lang.org/reference/conditional-compilation.html

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+// read_rdp_output reads incoming RDP bitmap frames from client at client_ref and forwards them to
+// handle_bitmap. handle_bitmap *must not* free the memory of CGOBitmap.
+#[no_mangle]
+pub extern "C" fn read_rdp_output(
```
Same goes here for the extern c wrapping style I commented about earlier.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+
+// read_rdp_output reads incoming RDP bitmap frames from client at client_ref and forwards them to
+// handle_bitmap. handle_bitmap *must not* free the memory of CGOBitmap.
+#[no_mangle]
```
isn't this implicit with extern "C", memory may serve me wrong though

xacrimon replied:
👍

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+    handle_bitmap: unsafe extern "C" fn(i64, CGOBitmap) -> CGOError,
+) -> CGOError {
+    let mut tcp_fd = 0;
+    if let Err(e) = with_client(&client_ref, |client| -> Result<(), ClientError> {
```
Generally closures don't need to define result types like this

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+    }
+    // Read incoming events.
+    // TODO: this doesn't always unblock after client.shutdown() was called. Figure out why.
+    while wait_for_fd(tcp_fd as usize) {
```
This uh seems like a loop that would gobble a lot of CPU as it's constantly making syscalls. Can you passively wait instead or is this not a problem?

xacrimon replied:
ah, clever 👍

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+                }
+                // These should never really be sent by the server to us.
+                RdpEvent::Pointer(_) => {
+                    println!("got unexpected pointer event from RDP server, ignoring");
```
You probably want to use a library like `log` for this.

xacrimon commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,359 @@
+                }
+            })?;
+            if !err.is_null() {
+                Err(ClientError::Other(format!(
```
Ideally you should define explicit enum variants with conversions and whatnot for these errors and not use weakly typed strings.

timothyb89 replied:
Maybe https://docs.rs/err-derive/0.3.0/err_derive/?

xacrimon replied:
`thiserror` is generally the recommend derive crate for this. The downside is that all derive crates incur lots of compile time bloat and smaller amounts of runtime size bloat. For small applications like this I generally just spend the required 5 minutes writing them.

xacrimon reviewed (commented):
Did an initial pass

awly commented:
@fspmarshall @xacrimon thanks for the feedback on strings.
I was thinking about [the dangers of using C-style strings](https://nee.lv/2021/02/28/How-I-cut-GTA-Online-loading-times-by-70/) and thought that implementing something that resembles both Go and Rust strings (ptr + len) would be safer.
But clearly it's very dangerous. I'll use C-style strings, since we pass around short ones anyway.

Commit:
flag tweaks and updated readme for macos

fspmarshall commented:
@awly In theory, you're totally right, but memory management across FFI boundary is hard, and excellent tooling exists for C-like strings.  Plus, rust's `CStr` type does actually make a single call to `strlen` when initialized and then treat the `CStr` like a ptr + len string from that point forward, so its performance is still quite good in comparison to string handling in normal C.

Commit:
Switch to C-style strings between Go and Rust

Commit:
Use regular top-level C function instead ofthe cgo jump function

Commit:
Fix bitmap buffer memory leak

Commit:
Consistent naming for CGO types

Commit:
Pass rust object reference to Go

Commit:
Clean up FFI error string management

Commit:
Extract a thin C wrapper for read_rdp_output

Commit:
Use the log crate in rust

Commit:
Small rust cleanups

Commit:
Fix shellcheck nit in run.sh

xacrimon reviewed (approved):
Looks correct to me, just test with Valgrind also. Look for things like UAFs and unreleased memory and any races.

timothyb89 reviewed (commented):
What's the expected behavior for the test client? It logs me out on my Windows box so the session does start, but nothing ever renders. 

`xfreerdp` and rdp-rs's own `mstsc-rs` example both seem to work fine. Maybe there's a problem with my environment? There's some "GLOBAL" log entries, but `mstsc-rs` produces the same output and works.

Commit:
Fix RDP client memory release

Commit:
Allow Rust code to compile on any unix

awly commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,380 @@
+    }
+}
+
+#[cfg(target_os = "linux")]
```
nice catch, updated

awly commented:
@timothyb89 does the web page remain blank or does it show a black rectangle?

In some cases I get
```
WARN[0080] Failed reading RDP output frame: failed forwarding RDP bitmap frame: RdpError(RdpError { kind: Disconnect, message: "MCS: Disconnect Provider Ultimatum" })
```
which means that the Windows box for some reason kills the connection.
I have not figured out why that happens yet, but it does work after a few tries.
When it does happen, you can see the login spinner for a second and then a black screen.

If you see a blank page, check the JS console for any errors. Maybe it's a browser incompatibility thing.

timothyb89 commented:
> @timothyb89 does the web page remain blank or does it show a black rectangle?
> 
> In some cases I get
> 
> ```
> WARN[0080] Failed reading RDP output frame: failed forwarding RDP bitmap frame: RdpError(RdpError { kind: Disconnect, message: "MCS: Disconnect Provider Ultimatum" })
> ```
> 
> which means that the Windows box for some reason kills the connection.
> I have not figured out why that happens yet, but it does work after a few tries.
> When it does happen, you can see the login spinner for a second and then a black screen.
> 
> If you see a blank page, check the JS console for any errors. Maybe it's a browser incompatibility thing.

There's occasionally a border around the canvas (I think in Chrome), but it just remains white with no console messages. I just thought to check the websocket inspector in Chrome's devtools and it's showing a continuous stream of binary messages to and from the browser. It might just be a rendering problem? (Same issue in Chrome, Firefox, and Safari on my MacBook, as well as an ancient Chrome install on an old Linux laptop)

awly commented:
Note to self: it's likely caused by the different Windows version on the remote end.
The websocket bitmap messages are suspiciously small (~200 bytes), but it could be efficient PNG encoding.
Need to set up a Windows 10 machine and try for myself.

awly committed:
Force Alpha channel to 100% always

For some reason, on Windows 10 with bitmap decompression the Alpha
always ends up as 0. This makes everything transparent.

awly commented:
@timothyb89 try now.
For some reason, Windows 10 bitmaps end up with Alpha value of 0 (transparent), which is why nothing is shown.
IDK if this is a problem in the rust library or something else, but for now I just override Alpha to 255 (100% opaque) and it makes the image show up.

You will definitely notice that all animations are super slow, I'll have to work a lot on performance.

awly committed:
Implement screen size and credential negotiation

timothyb89 commented:
> @timothyb89 try now.
> For some reason, Windows 10 bitmaps end up with Alpha value of 0 (transparent), which is why nothing is shown.
> IDK if this is a problem in the rust library or something else, but for now I just override Alpha to 255 (100% opaque) and it makes the image show up.
> 
> You will definitely notice that all animations are super slow, I'll have to work a lot on performance.

That did the trick! I can see what you mean about performance, but it does work pretty well now.

Commit:
desktop protocol: remove password prompt

It was decided that supporting passwords natively is a bad product
decision. We will only support certificate-based authn and only in
ActiveDirectory environments.

Until we implement certificate support, passwords can be injected via an
environment variable for testing. It will be removed before the beta
release.

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+// Package deskproto implements the desktop protocol encoder/decoder.
```
Nit: License header missing.

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+	TypePNGFrame         = MessageType(2)
+	TypeMouseMove        = MessageType(3)
+	TypeMouseButton      = MessageType(4)
+	TypeKeyboardButton   = MessageType(5)
+	TypeClientUsername   = MessageType(7)
```
Where is `MessageType(6)`? :)

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+type MessageType byte
+
+const (
+	TypeClientScreenSpec = MessageType(1)
```
Might be helpful to add godocs on these to explain what they represent. Some are self-explanatory, some are not.

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+		return nil, trace.BadParameter("input desktop protocol message is empty")
+	}
+	switch buf[0] {
+	case byte(TypeClientScreenSpec):
```
Nit: If you did `switch MessageType(buf[0])` instead, you wouldn't need byte casts in each case branch.

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+	if err := binary.Write(buf, binary.BigEndian, header{
+		Type:   byte(TypePNGFrame),
+		Left:   uint32(f.Img.Bounds().Min.X),
+		Top:    uint32(f.Img.Bounds().Min.Y),
```
Shouldn't "top" be Max.Y and "bottom" - Min.Y?

awly replied:
Desktop coordinate systems typically go from top-left (min) to bottom-right (max).
I think that mimics the order of how screens light up the pixels.

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+	// PNG encoding does not preserve offset image bounds.
+	// Opportunistically restore them based on the header.
+	switch img := img.(type) {
+	case *image.RGBA:
+		img.Rect = image.Rect(int(header.Left), int(header.Top), int(header.Right), int(header.Bottom))
+	case *image.NRGBA:
+		img.Rect = image.Rect(int(header.Left), int(header.Top), int(header.Right), int(header.Bottom))
```
Feels like these two can be combined in a single `case *image.RGBA, *image.NRGB`.

awly replied:
Not in a type switch. Otherwise, `img` could be one of two types and the compiler doesn't allow it.

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+
+// KeyboardButton is the keyboard button press message.
+// https://github.com/gravitational/teleport/blob/master/rfd/0037-desktop-access-protocol.md#4---keyboard-input
+type KeyboardButton struct {
```
I'm curious - how are multi-key combinations (i.e. `Ctrl-Alt-Del`) are represented in the protocol?

awly replied:
As one message per key press - separate `ctrl`/`alt`/`del`.
Each message represents key going up or down only, so the server can track which keys are currently down and interpret those as a combination.

r0mant commented on lib/srv/desktop/deskproto/proto.go:
```diff
@@ -0,0 +1,254 @@
+	return s, trace.Wrap(err)
+}
+
+// ClientUsername is the client username.
```
Might be helpful to explain in which case this message is sent as well.

awly replied:
Message meaning and sequencing is explained in the RFD.
I'd rather keep that in one central place and link to it.

r0mant commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -0,0 +1,345 @@
+
+	clientWidth, clientHeight uint16
+	username                  string
+	rustClientRef             *C.Client
```
Nit: `Ref` suffix seems superfluous, I'd consider just naming this `rustClient` (unless it's intentional).

r0mant commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -0,0 +1,345 @@
+}
+
+// Options for creating a new Client.
+type Options struct {
```
Would it make sense to name this `Config` to be consistent with our naming everywhere else?

r0mant commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -0,0 +1,345 @@
+		defer unregisterClient(clientRef)
+
+		if err := cgoError(C.read_rdp_output(c.rustClientRef, C.int64_t(clientRef))); err != nil {
+			logrus.Warningf("Failed reading RDP output frame: %v", err)
```
Nit: I'd consider making logger a part of the `Client` with proper component set and do `c.log` everywhere.

r0mant commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -0,0 +1,345 @@
+func unregisterClient(i int64) {
+	clientsMu.Lock()
+	defer clientsMu.Unlock()
+	delete(clients, i)
```
Does this also need to decrement `clientsIndex`?

awly replied:
`clientsIndex` is like a sequential "ID" for clients, not reusable

fspmarshall commented on lib/srv/desktop/rdp/rdpclient/src/lib.rs:
```diff
@@ -0,0 +1,382 @@
+    tcp_fd: usize,
+}
+
+impl Client {
+    fn into_raw(self) -> *mut Client {
+        Box::into_raw(Box::new(self))
+    }
+    unsafe fn from_raw<'a>(ptr: *mut Client) -> &'a mut Client {
+        ptr.as_mut().unwrap()
+    }
+    unsafe fn free(&mut self) {
+        Box::from_raw(self as *mut Client);
+    }
+}
```
Oh, and its best not to use unwraps in code like this.  Panicing across the FFI boundary is undefined behavior.

awly replied:
Thanks, done!

Commit:
Address review feedback
